### PR TITLE
Reduce extent

### PR DESCRIPTION
**Pull request by jreadey**
Enable shape reduction and handle proper re-initialization, deletion of affected chunks.

Includes updates for:
  * #271 
  * #272 
  * #276

### CODE REVIEW

File: hsds/chunk_crawl.py [hunk 1]
```diff
@@ -43,6 +45,33 @@
...
+
+    fill_value = None
+    type_json = dset_json["type"]
+    dt = createDataType(type_json)
+
+    if "creationProperties" in dset_json:
+        cprops = dset_json["creationProperties"]
+        if "fillValue" in cprops:
+            fill_value_prop = cprops["fillValue"]
+            log.debug(f"got fo;;+value_prop: {fill_value_prop}")
```
Comment (mattjala) [inline hsds/chunk_crawl.py:63]: Typo: `got fo;;+value_prop` -> `got fill_value_prop`

File: hsds/chunk_sn.py [hunk 2]
```diff
@@ -87,35 +88,14 @@ async def get_slices(app, select, dset_json, bucket=None):
...
-    maxdims = getDsetMaxDims(dset_json)
-
-    # refetch the dims if the dataset is extensible and request or hasn't
-    # provided an explicit region
-    if isExtensible(dims, maxdims) and (select is None or not select):
-        kwargs = {"bucket": bucket, "refresh": True}
-        dset_json = await getObjectJson(app, dset_id, **kwargs)
-        dims = getShapeDims(dset_json["shape"])
-
-    slices = None  # selection for read
```
Comment (mattjala) [inline hsds/chunk_sn.py:99]: Should `slices` still be declared before the `try` block in the new version of this function?

File: hsds/chunk_sn.py [hunk 3]
```diff
@@ -924,15 +958,14 @@ async def PUT_Value(request):
...
-                )
+                msg = f"crawler failed for page: {page_number+1} with status: {crawler_status}"
+                log.warn(msg)
             else:
                 log.info("crawler write_chunk_hyperslab successful")
 
     else:
         #
-        # Do point PUT
+        # Do point post
```
Comment (mattjala) [inline hsds/chunk_sn.py:968]: Other log messages statements here indicate this is a PUT request. If this is a POST request when done for points, I think the following log messages should be updated as well.

File: hsds/dset_lib.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    else:
+        arr = np.zeros([1,], dtype=dt, order="C")
+
+    return arr
+
+
+async def removeChunks(app, chunk_ids, bucket=None):
+    """ Remove chunks with the given ids """
+
+    # this should only be called from a SN
```
Comment (mattjala) [inline hsds/dset_lib.py:58]: Would it make more sense for this function to be in `servicenode_lib.py`?

File: hsds/dset_lib.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+            continue
+        chunk_ids.append(chunk_id)
+
+    log.debug(f"getAllocattedChunkIds - got {len(chunk_ids)} ids")
+    return chunk_ids
+
+
+async def reduceShape(app, dset_json, shape_update, bucket=None):
+    """ Given an existing dataset and a new shape,
+        Reinitialize and edge chunks and delete any chunks
```
Comment (mattjala) [inline hsds/dset_lib.py:162]: Typo: `Reinitialize and edge chunks` -> `Reinitialize any edge chunks`

File: hsds/dset_lib.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        for i in range(rank):
+            if chunk_coord[i] + layout[i] > shape_update[i]:
+                skip = False
+                break
+        if skip:
+            log.debug(f"chunk_id {chunk_id} no action needed")
+            continue
+
+        reinit = False
+        for n in range(rank):
```
Comment (mattjala) [inline hsds/dset_lib.py:210]: I think this could be done with better performance via `reinit = np.any(chunk_coord[:rank] < shape_update[:rank])`

**Review by mattjala — APPROVED** [review]

File: hsds/chunk_crawl.py (hunk 1 above)
Comment (jreadey) [reply hsds/chunk_crawl.py]: fixed

File: hsds/chunk_sn.py (hunk 2 above)
Comment (jreadey) [reply hsds/chunk_sn.py:99]: updated

File: hsds/chunk_sn.py (hunk 3 above)
Comment (jreadey) [reply hsds/chunk_sn.py]: updated

File: hsds/dset_lib.py (hunk 4 above)
Comment (jreadey) [reply hsds/dset_lib.py:595]: That would be fine but looking to put all dataset specific stuff in dset_lib.py

File: hsds/dset_lib.py (hunk 5 above)
Comment (jreadey) [reply hsds/dset_lib.py]: updated

File: hsds/dset_lib.py (hunk 6 above)
Comment (jreadey) [reply hsds/dset_lib.py]: not sure how much better perorming it will be, but your method is more readable.  changed.

**Review by mattjala — APPROVED** [review]

**Pull request merged**